Bluetooth: audio: Improve stream coupling for CIS as the unicast client

MariuszSkamra opened the pull request:
Allow  the streams to be paired when creating unicast group. This will
allow to reuse the same ISO for the paired streams.

Fixes: #51796
Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: Improve stream coupling for CIS as the unicast client

Allow  the streams to be paired when creating unicast group. This will
allow to reuse the same ISO for the paired streams.

Fixes: #51796
Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: bsim: audio: Create unicast group prior stream configuration

This updates the test to create the unicast group with non-configured
streams.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -1135,10 +1133,10 @@ int bt_audio_unicast_group_add_streams(struct bt_audio_unicast_group *unicast_gr
-					       params[num_added].stream,
-					       params[num_added].qos,
-					       params[num_added].dir);
+		err = unicast_group_add_stream_pair(unicast_group,
+					       params[num_added].in,
+					       params[num_added].out,
+					       params[num_added].qos);
```
```suggestion
		err = unicast_group_add_stream_pair(unicast_group,
					            params[num_added].in,
					            params[num_added].out,
					            params[num_added].qos);
```

Thalley commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -859,22 +830,48 @@ static void unicast_client_codec_qos_to_iso_qos(struct bt_audio_iso *iso,
+	sys_slist_append(&group->streams, &stream->_node);
+}
+
+static int unicast_group_add_stream_pair(struct bt_audio_unicast_group *group,
+					 struct bt_audio_stream *in,
+					 struct bt_audio_stream *out,
+					 struct bt_codec_qos *qos)
```
Given that the last 3 arguments are always in the `bt_audio_unicast_group_stream_param`, we could just do
```suggestion
static int unicast_group_add_stream_pair(struct bt_audio_unicast_group *group,
					 struct bt_audio_unicast_group_stream_param *param)
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c:
```diff
@@ -398,12 +398,12 @@ static void test_main(void)
 	for (unsigned int i = 0U; i < iterations; i++) {
 		printk("\n########### Running iteration #%u\n\n", i);
 
+		printk("Creating unicast group\n");
+		create_unicast_group(&unicast_group);
```
Keep in mind that until `CONFIG_BT_CTLR_CENTRAL_ISO` gets enabled, creating the group here doesn't do anything. 

Have you verified your changes by running the sample?

MariuszSkamra replied:
Yes, I run this PR with https://github.com/zephyrproject-rtos/zephyr/pull/53450 fix

MariuszSkamra commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1951,18 +1951,14 @@ int bt_audio_stream_send(struct bt_audio_stream *stream, struct net_buf *buf,
-	struct bt_audio_stream *stream;
+	/** Pointer to a input stream object. */
+	struct bt_audio_stream *in;
 
-	/** The QoS settings for the @ref bt_audio_unicast_group_stream_param.stream. */
-	struct bt_codec_qos *qos;
+	/** Pointer to a output stream object. */
+	struct bt_audio_stream *out;
```
Sure, I don't mind using different naming :)

MariuszSkamra commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -295,7 +293,7 @@ static void audio_timer_timeout(struct k_work *work)
 	 */
 	for (size_t i = 0U; i < configured_sink_stream_count; i++) {
-		struct bt_audio_stream *stream = &streams[i];
+		struct bt_audio_stream *stream = &sinks[i].stream;
```
I agree, maybe it would be better to name it `tx` and `rx` respectively?

Thalley replied:
> I agree, maybe it would be better to name it `tx` and `rx` respectively?

I think so :) 

Or c_to_s (client to server) and s_to_c if you want to use an absolute reference, instead of relative, but TX/RX is probably best

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1951,18 +1951,14 @@ int bt_audio_stream_send(struct bt_audio_stream *stream, struct net_buf *buf,
-	 * different directions, they may use the same CIS.
-	 */
-	enum bt_audio_dir dir;
+	/** The QoS settings for the stream objects. */
+	struct bt_codec_qos *qos;
```
As per the Core Spec, the SDU, PHY and RTN may also be different: 
![image](https://user-images.githubusercontent.com/2520413/211305188-e7b1f70f-6a16-4800-b2f4-c060616f18d4.png)


So basically we have

- Packing shall be the same for all CIS
- Framing may actually be different (depending on which core spec version is used; in 5.3 you can actually have different framing for different CIS, but will always be the same for each direction)
- Max Transport Latency shall be the same for all CIS in a direction, but may be different in each direction
- SDU, PHY and RTN may be different for each CIS and for each direction

MariuszSkamra commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -51,21 +53,17 @@ static K_SEM_DEFINE(sem_security_updated, 0, 1);
 static K_SEM_DEFINE(sem_sources_discovered, 0, 1);
 static K_SEM_DEFINE(sem_stream_configured, 0, 1);
-static K_SEM_DEFINE(sem_stream_qos, 0, 1);
+static K_FIFO_DEFINE(fifo_stream_qos);
```
Do you mean calling `k_sem_init(sem, initial_count, limit)` before QoS instead? I was unsure it's allowed to call this API function multiple times on the same semaphore object.

Thalley replied:
I was more thinking about doing
```c
static K_SEM_DEFINE(sem_stream_qos, 0, ARRAY_SIZE(sinks) + ARRAY_SIZE(sinks));
```

and then where you do 
```c
k_fifo_put(&fifo_stream_qos, stream);
``` 
you just revert to 
```c
k_sem_give(&sem_stream_qos);
```

and where you do 
```c
k_fifo_get(&fifo_stream_qos, K_FOREVER);
```
you just do

```c
k_sem_take(sem_stream_qos);
```

It will have the same effect as the FIFO, but just using a semaphore instead. Whether or not it is better, is a good question :)

MariuszSkamra commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -990,12 +975,34 @@ static int enable_streams(void)
+		init_lc3();
+	}
+
+	err = enable_app_streams(sinks, configured_sink_stream_count);
+	if (err != 0) {
+		return err;
+	}
+
+	err = enable_app_streams(sources, configured_source_stream_count);
+	if (err != 0) {
+		return err;
+	}
```
I did it to avoid code duplicates

Thalley replied:
Ah, that makes sense :)

MariuszSkamra commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -1048,6 +1072,13 @@ void main(void)
 	while (true) {
 		reset_data();
 
+		printk("Creating unicast group\n");
+		err = create_group();
+		if (err != 0) {
+			return;
+		}
+		printk("Unicast group created\n");
```
The sample removes the group after device disconnection. I wanted to keep the same logic

MariuszSkamra commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -1001,18 +990,22 @@ int bt_audio_unicast_group_create(struct bt_audio_unicast_group_param *param,
-			LOG_DBG("params[%zu] stream (%p) already part of group %p",
-				i, stream_param->stream,
-				stream_param->stream->group);
+		if (stream_param->in && stream_param->in->group != NULL) {
```
I just moved the code, but I'll change it as you suggested

MariuszSkamra committed:
samples: unicast_audio_client: Fix enabling the streams

This fixes the application that was not waiting for QoS configuration of
all streams in the group. The QoS operation is done on group of streams,
while the status of each one of the streams is received separately.
Meaning that the application has to control which streams were already
configured before enabling them.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -2020,7 +2021,7 @@ int bt_audio_unicast_group_create(struct bt_audio_unicast_group_param *param,
  *  @return 0 in case of success or negative value in case of error.
  */
 int bt_audio_unicast_group_add_streams(struct bt_audio_unicast_group *unicast_group,
-				       struct bt_audio_unicast_group_stream_param params[],
+				       struct bt_audio_unicast_group_stream_pair_param params[],
 				       size_t num_param);
```
Should this be renamed to `add_stream_pairs`?

MariuszSkamra replied:
Not necessarily, as user might not pair the streams.

Thalley commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -954,10 +956,9 @@ static int set_stream_qos(void)
-	if (err != 0) {
-		printk("failed to take sem_stream_qos (err %d)\n", err);
-		return err;
+	for (size_t i = configured_stream_count; i > 0; i--) {
```
```suggestion
	for (size_t i = 0U; i < configured_stream_count; i++) {
```

Thalley commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -975,6 +984,84 @@ static void unicast_group_free(struct bt_audio_unicast_group *group)
+}
+
+static int group_qos_common_set(const struct bt_codec_qos **group_qos,
+				struct bt_audio_unicast_group_stream_pair_param *param)
```
```suggestion
				const struct bt_audio_unicast_group_stream_pair_param *param)
```

Thalley commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -975,6 +984,84 @@ static void unicast_group_free(struct bt_audio_unicast_group *group)
+	return 0;
+}
+
+static int stream_pair_param_check(struct bt_audio_unicast_group_stream_pair_param *param)
```
```suggestion
static int stream_pair_param_check(const struct bt_audio_unicast_group_stream_pair_param *param)
```

Thalley commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -975,6 +984,84 @@ static void unicast_group_free(struct bt_audio_unicast_group *group)
 	group->allocated = false;
 }
 
+static int stream_param_check(struct bt_audio_unicast_group_stream_param *param)
```
```suggestion
static int stream_param_check(const struct bt_audio_unicast_group_stream_param *param)
```

Thalley reviewed (changes requested):
Some minor comments, otherwise LGTM

fredrikdanebjer commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -975,6 +984,84 @@ static void unicast_group_free(struct bt_audio_unicast_group *group)
+	if (param->rx_param != NULL) {
+		if (*group_qos == NULL) {
+			*group_qos = param->rx_param->qos;
+		} else if (!unicast_group_valid_qos(*group_qos, param->rx_param->qos)) {
```
This function was removed in merged https://github.com/zephyrproject-rtos/zephyr/pull/53119

fredrikdanebjer commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -900,26 +882,46 @@ static int configure_streams(void)
+
+		if (i < configured_sink_stream_count) {
+			stream_param = &stream_params[i];
+			pair_params[i].tx_param = stream_param;
```
```suggestion
			pair_params[i].tx_param = &stream_params[i];
```
and subsequently removve the `struct bt_audio_unicast_group_stream_param *stream_param;`

fredrikdanebjer reviewed (approved):
lgtm
